test/extensions/filters/http/jwt_authn/group_verifier_test.cc

```diff
@@ -0,0 +1,537 @@
+  mock_auths_.push_back(std::move(mock_auth));
+  for (int i = 0; i < 2; ++i) {
+    mock_auth = std::make_unique<MockAuthenticator>();
+    EXPECT_CALL(*mock_auth.get(), doVerify(_, _, _))
+        .WillRepeatedly(Invoke([&](Http::HeaderMap&, std::vector<JwtLocationConstPtr>*,
+                                   AuthenticatorCallback callback) { callback(Status::Ok); }));
+    mock_auths_.push_back(std::move(mock_auth));
+  }
+  createVerifier();
+  EXPECT_CALL(mock_cb_, onComplete(_)).WillOnce(Invoke([](const Status& status) {
+    ASSERT_EQ(status, Status::Ok);
+  }));
+  auto headers = Http::TestHeaderMapImpl{};
+  context_ = Verifier::createContext(headers, &mock_cb_);
+  verifier_->verify(context_);
+}
+
+// Test contains a 2 provider_name in a require any along with another provider_name in require all.
+// Test simulates all require any OK and proivder_name is OK.
+TEST_F(GroupVerifierTest, TestAnyInAllBothInRequireAnyIsOk) {
+  MessageUtil::loadFromYaml(AllWithAny, proto_config_);
+  std::vector<AuthenticatorCallback> callbacks;
+  for (int i = 0; i < 3; ++i) {
+    auto mock_auth = std::make_unique<MockAuthenticator>();
+    EXPECT_CALL(*mock_auth.get(), doVerify(_, _, _))
+        .WillRepeatedly(Invoke(
+            [&](Http::HeaderMap&, std::vector<JwtLocationConstPtr>*,
+                AuthenticatorCallback callback) { callbacks.push_back(std::move(callback)); }));
+    mock_auths_.push_back(std::move(mock_auth));
+  }
+  createVerifier();
+  EXPECT_CALL(mock_cb_, onComplete(_)).WillOnce(Invoke([](const Status& status) {
+    ASSERT_EQ(status, Status::Ok);
+  }));
+  auto headers = Http::TestHeaderMapImpl{};
+  context_ = Verifier::createContext(headers, &mock_cb_);
+  verifier_->verify(context_);
+  for (const auto callback : callbacks) {
+    callback(Status::Ok);
+  }
+}
+
+// Test contains a 2 provider_name in a require any along with another provider_name in require all.
+// Test simulates all require any failed and proivder_name is OK.
+TEST_F(GroupVerifierTest, TestAnyInAllBothInRequireAnyFailed) {
+  MessageUtil::loadFromYaml(AllWithAny, proto_config_);
+  std::vector<AuthenticatorCallback> callbacks;
+  for (int i = 0; i < 3; ++i) {
+    auto mock_auth = std::make_unique<MockAuthenticator>();
+    EXPECT_CALL(*mock_auth.get(), doVerify(_, _, _))
+        .WillRepeatedly(Invoke(
+            [&](Http::HeaderMap&, std::vector<JwtLocationConstPtr>*,
+                AuthenticatorCallback callback) { callbacks.push_back(std::move(callback)); }));
+    mock_auths_.push_back(std::move(mock_auth));
+  }
+  createVerifier();
+  EXPECT_CALL(mock_cb_, onComplete(_)).WillOnce(Invoke([](const Status& status) {
+    ASSERT_EQ(status, Status::JwksFetchFail);
+  }));
+  auto headers = Http::TestHeaderMapImpl{};
+  context_ = Verifier::createContext(headers, &mock_cb_);
+  verifier_->verify(context_);
+  callbacks[0](Status::JwksFetchFail);
+  callbacks[1](Status::JwksFetchFail);
```

Comment:
maybe callback_ can be a map with key as issuer and value is callback.  Then the test has more control over which callback is calling.